src/types/Stream.d.ts

```diff
@@ -0,0 +1,53 @@
+import Client from './Client'
+import { ExprArg } from './query'
+import { values } from './values'
+
+export type StreamSubscription =
+  | StreamStartSubscription
+  | StreamVersionSubscription
+  | StreamErrorSubscription
+  | StreamHistoryRewriteSubscription
+
+interface Subscription {
+  start: () => this
+  close: () => void
+  txnTS: number
+}
```

Comment:
The subscription itself does not contain data. The previews version had the correct type signature:

```js
interface Subscription {
  on: (type: SubscriptionEvents, callback: Function) => this
  start: () => this
  close: () => this
}
```

What I wanted to try is to tie the `SubscriptionEvents` literal string with the type of the `callback` parameter.

For example:

```js
client.stream(ref).on('start', ts => ...)
//                             ^ 
// I wanted typescript to understand that this is a FaunaTime
```

@BrunoQuaresma mentioned earlier in the project that maybe that's possible. I don't know much about TS, though. I'm okay if that's not feasible.

Reply:
https://github.com/fauna/faunadb-js/pull/356#issuecomment-713202337